src/textual/widgets/_data_table.py

```diff
@@ -2127,11 +2132,19 @@ def sort_by_column_keys(
             result = itemgetter(*columns)(row_data)
             return result
 
+        _key = key
+        if key and columns:
+
+            def _key(row):
```

Comment:
No, it's not necessary, I just included it in case someone provided multiple columns to their sort function and wanted to easily access them via the `RowKey`. Probably overkill, so I will just provide the cell data.

Reply:
I'm not sure I follow... The original code provided a tuple of `tuple[RowKey, dict[ColumnKey | str, CellType]]` to the default `sort_by_column_keys()` function, then after sorting, it uses the `RowKey` part of the now sorted tuples to update the `_row_locations` dictionary. I just followed that same logic when adding the catch for `key` and `columns`. I could re-write the catch to be more clear and in line with the default `sort_by_column_keys()` function like below.

```python
if key and columns:

    def _key(row):
        _, row_data = row
        return key(itemgetter(*columns)(row_data))
```

Let me know if I am missing what you are point out?

Reply:
I see how the `row_key` is used now. The key function gets a tuple of row key, and row data, but it discards the row key (which you unpack as `_`).

I think it would be surprising to have the row key there (and its not documented in the signature). Could we allow the key function to accept just the row data? That's what I would expect to be the default.

I guess you would need to wrap the key function. Something like the following:

```python
def key_wrapper(row):
    row_key, row_data = row
    return (row_key, key(row_data))
```

Reply:
Sorry, can I back up a bit and get your thoughts to make sure we are on the same page... I think we are working toward the same thing since I am already "wrapping" the key function to only send the row_data to the original user-supplied `key` function...

First things first. I should have asked this in the beginning, I guess. The `_data_table.sort()` API in `main` raises an exception if no `columns` are passed since the base `sort_by_column_keys()` function uses `itemgetter` to get each item from `columns`. When a `key` is provided should the same logic be followed, or should all row data be sent to the provided `key` function?

If the same exception should be raised if no `columns` are provided, then this "wraps" the user-supplied `key` function and only sends the actual row data.

```python
_key = key
if key:

    def _key(row: tuple[RowKey, dict[ColumnKey | str, CellType]]) -> Any:
        _, row_data = row
        return key(itemgetter(*columns)(row_data))
```

If all data should be sent in the absence of `columns`, then a ternary to send the correct data should do the trick.

```python
_key = key
if key:

    def _key(row):
        _, row_data = row
        return key(itemgetter(*columns)(row_data) if columns else tuple(row_data.values()))
```

In both cases, only the actual row_data is being passed to the user-supplied `key` function, the only reason for the `row_key` tagging along is to be able to update the `TwoWayDict` the API has set up to track all of the data. To my knowledge, if you don't send along with the `RowKey` to the built-in `sorted` method, you will have to do more work to reconstruct the `self._row_locations` dict after sorting.

```python
ordered_rows = sorted(
    self._data.items(),
    key=_key if key is not None else sort_by_column_keys,
    reverse=reverse,
)
self._row_locations = TwoWayDict(
    {row_key: new_index for new_index, (row_key, _) in enumerate(ordered_rows)}
)
```